### PR TITLE
Subscribe by tags - issue #143

### PR DESCRIPTION
**Pull request by schandrika**
Added subscribe and publish by tags feature to pubsub subsystem. Issue #143

### CODE REVIEW

**Review by shwethanidd — CHANGES_REQUESTED** [review]

File: src/volttron/client/vip/agent/subsystems/pubsub.py [hunk 1]
```diff
@@ -513,16 +664,49 @@ def publish(self, peer: str, topic: str, headers=None, message=None, bus=""):
...
+         given tag condition. defaults to "devices"
+        :type topic_source str
+        :param headers: header info for the message
+        :type headers: None or dict
+        :param message: actual message
+        :type message: None or any
+        :param bus: bus
+        :type bus: str
+        :param publish_multiple: Boolean value if publish can be done to multiple topics if tag_condition matches multiple topics
+        :type publish_multiple: boolean
```
[the remark this answers is not in the document]
Comment (schandrika) [reply src/volttron/client/vip/agent/subsystems/pubsub.py]: Updated publish_multiple to max_publish_count parameter

**Comment by schandrika** [comment]
Updates to be done based on review discussion: - 
1. move topic_source as the last parameter in publish/subscribe by tags method
2. Remove the parameter 'peer' (its always 'pubsub') and parameter 'bus' (always used with default value) from publish and subscribe methods
3. Move 'persistent_queue' parameter as part of kwargs as this is applicable only for rmq and not zmq

**Comment by schandrika** [comment]
Completed 1, 3, and 4 above. Will create a new PR for point number 2 above as that will involve updating and testing all agent code and tests

File: src/volttron/client/vip/agent/subsystems/pubsub.py [hunk 2]
```diff
@@ -131,120 +146,86 @@ def _process_callback(self, sender, bus, topic, headers, message):
...
-        return len(subscribers)
+    @spawn
+    def refresh_tag_subscriptions(self):
+        def platform_subscriptions():
+            return defaultdict(subscriptions)
+
+        def subscriptions():
+            return defaultdict(set)
+
+        print("refreshing tags")
```
Comment (craig8) [inline src/volttron/client/vip/agent/subsystems/pubsub.py:189]: remove the print make it a log if you must.

**Review by craig8 — CHANGES_REQUESTED** [review]

**Review by craig8 — APPROVED** [review]

**Pull request merged**